File: tests/cpp/test_database.cpp
#include <gtest/gtest.h>
#include "mollerdb/Database.h"
#include <stdexcept>

// Test fixture for Database tests
class DatabaseTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Setup code if needed
    }

    void TearDown() override {
        // Cleanup code if needed
    }
};

// Test that Database constructor accepts a connection string
TEST_F(DatabaseTest, ConstructorWithConnectionString) {
    // This test verifies that the Database class can be constructed with a connection string
    // Note: This will throw if the database is not accessible, which is expected in a CI environment
    // We're testing the parsing logic here
    EXPECT_THROW({
        moller::db::Database db("host=localhost dbname=test user=test password=test");
    }, std::runtime_error);
}

// Test that Database constructor throws on connection failure
TEST_F(DatabaseTest, ConstructorThrowsOnInvalidConnection) {
    // Test with an invalid connection string
    EXPECT_THROW({
        moller::db::Database db("host=invalid_host dbname=nonexistent user=nobody password=wrong");
    }, std::runtime_error);
}

// Test connection string parsing with various formats
TEST_F(DatabaseTest, ConnectionStringParsing) {
    // Test various connection string formats
    // These should all throw since we can't actually connect, but we're verifying the parsing doesn't crash
    
    // Standard format
    EXPECT_THROW({
        moller::db::Database db("host=localhost dbname=testdb user=testuser password=testpass port=5432");
    }, std::runtime_error);
    
    // Minimal format
    EXPECT_THROW({
        moller::db::Database db("host=localhost dbname=testdb");
    }, std::runtime_error);
}

// Test that get_main_detector_results_arrow returns nullptr (placeholder implementation)
TEST_F(DatabaseTest, GetMainDetectorResultsReturnsNullptr) {
    // Since we can't connect to a real database in tests, we can only test the placeholder behavior
    // This test will need to be updated when real implementation is added
    
    // We expect construction to fail, so we can't test the method directly yet
    // This is a placeholder for future implementation
    EXPECT_TRUE(true);  // Placeholder assertion
}

// Test invalid port in connection string
TEST_F(DatabaseTest, InvalidPortThrowsException) {
    EXPECT_THROW({
        moller::db::Database db("host=localhost dbname=test user=test password=test port=invalid");
    }, std::invalid_argument);
}

// Main function for the test runner
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/python/test_database.py
"""
Tests for the mollerdb Database class.

Tests Database class instantiation, methods, and error handling.
"""

import pytest


def test_database_constructor_with_connection_string(mock_connection_string):
    """Test that Database can be instantiated with a connection string."""
    import mollerdb
    
    # This will throw a RuntimeError since we can't connect to a real database
    # We're testing that the constructor accepts the string and attempts connection
    with pytest.raises(RuntimeError):
        db = mollerdb.Database(mock_connection_string)


def test_database_constructor_with_invalid_connection(invalid_connection_string):
    """Test that Database throws an error with an invalid connection string."""
    import mollerdb
    
    # Should raise RuntimeError for connection failure
    with pytest.raises(RuntimeError):
        db = mollerdb.Database(invalid_connection_string)


def test_database_has_get_main_detector_results_method():
    """Test that the Database class has the get_main_detector_results method."""
    import mollerdb
    
    # Check that the method exists (even though we can't call it without a connection)
    assert hasattr(mollerdb.Database, 'get_main_detector_results')


def test_database_constructor_with_invalid_port():
    """Test that Database throws an error with an invalid port."""
    import mollerdb
    
    # Should raise an exception for invalid port
    with pytest.raises(Exception):  # Could be ValueError or RuntimeError
        db = mollerdb.Database("host=localhost dbname=test user=test password=test port=invalid")
